dvc/remote/base.py

```diff
@@ -787,6 +779,15 @@ def _checkout(
             path_info, checksum, force, progress_callback=progress_callback
         )
 
+    def get_files_number(self, checksum):
```

Comment:
Let's use this in dvc/output/base.py:get_files_number, instead of duplicating :slightly_smiling_face:

Reply:
The problem is `output.base` has `checksum` and `use_cache` (as well as `is_dir_checksum` and `dir_cache`) as member variables/properties.

Reply:
@casperdcl True, but we could
```
def get_files_number(self):
    if not self.use_cache:
        return 0

    return self.cache.get_files_number(self.checksum)
```

Reply:
but `self.use_cache` isn't defined

Reply:
also `OutputBase` and `RemoteBASE` don't seem to inherit from each other (unless there's some patching going on) so don't see how to avoid duplication unless they're made to inherit from the same base class

Reply:
@casperdcl I'm saying that OutputBase.get_files_number could look like
```
def get_files_number(self):
    if not self.use_cache:
        return 0

    return self.cache.get_files_number(self.checksum)
```
RemoteBase.get_files_number is self.cache.get_files_number for it.

Reply:
right but now (9b73ff72) there's another test failing (`TestGetFilesNumber.test_return_mutiple_for_dir`) where it expected `get_files_number() == 2` not `0`

Reply:
@efiop was waiting for a reply here just in case you missed it

Reply:
@casperdcl I totally did, thanks for another ping! Yes, I see. It is failing because it is patching `dir_cache` of the output, and now we need to patch `RemoteBASE.get_dir_cache`

Reply:
@casperdcl basically
```
diff --git a/tests/unit/output/test_local.py b/tests/unit/output/test_local.py 
index 8149e83f..6e2af113 100644                                                
--- a/tests/unit/output/test_local.py                                          
+++ b/tests/unit/output/test_local.py                                          
@@ -2,6 +2,7 @@ from mock import patch                                         
                                                                               
 from dvc.stage import Stage                                                   
 from dvc.output import OutputLOCAL                                            
+from dvc.remote.local import RemoteLOCAL                                      
                                                                               
 from tests.basic_env import TestDvc                                           
                                                                               
@@ -32,8 +33,12 @@ class TestGetFilesNumber(TestDvc):                          
         self.assertEqual(0, o.get_files_number())                             
                                                                               
     @patch.object(OutputLOCAL, "checksum", "12345678.dir")                    
-    @patch.object(OutputLOCAL, "dir_cache", [{"md5": "asdf"}, {"md5": "qwe"}])
-    def test_return_mutiple_for_dir(self):                                    
+    @patch.object(                                                            
+        RemoteLOCAL,                                                          
+        "get_dir_cache",                                                      
+        return_value=[{"md5": "asdf"}, {"md5": "qwe"}],                       
+    )                                                                         
+    def test_return_mutiple_for_dir(self, mock_get_dir_cache):                
         o = self._get_output()                                                
                                                                               
         self.assertEqual(2, o.get_files_number())                             
```

should do the trick :slightly_smiling_face:

Reply:
yup that was exactly the sort of thing I was thinking of but didn't want to spend time trying to fix it myself.